fastdtw/_fastdtw.pyx

```diff
@@ -0,0 +1,166 @@
+#!python
+#cython: boundscheck=False, cdivision=True, wraparound=False
+
+from __future__ import absolute_import, division
+
+from libcpp.vector cimport vector  # noqa
+from libcpp.map cimport map
+from libcpp.set cimport set
+from libcpp.utility cimport pair
+from cython.operator cimport dereference as deref
+from cython.operator cimport preincrement
+
+cdef struct CostEntry:
+    double cost
+    int prev_x_idx
+    int prev_y_idx
+
+
+def fastdtw(x, y, radius=1, dist=lambda a, b: abs(a - b)):
+    min_time_size = radius + 2
+
+    if len(x) < min_time_size or len(y) < min_time_size:
+        return dtw(x, y, dist=dist)
+
+    x_shrinked = __reduce_by_half(x)
+    y_shrinked = __reduce_by_half(y)
+    distance, path = fastdtw(x_shrinked, y_shrinked, radius=radius, dist=dist)
+    window = __expand_window(path, len(x), len(y), radius)
+    return dtw(x, y, window, dist=dist)
+
+
+def dtw(x, y, window=None, dist=lambda a, b: abs(a - b)):
+    cdef unsigned int len_x, len_y, i, j, k, len_window
+
+    cdef double inf, cost_up, cost_left, cost_cnr, dt
+    cdef pair[int, int] key_up, key_left, key_cnr, key_next
+    cdef vector[pair[int, int]] window_vec
+    len_x, len_y = len(x), len(y)
+    if window is None:
+        window = [(i, j) for i in range(len_x) for j in range(len_y)]
+
+    len_window = len(window)
+    window_vec.resize(len_window)
+    for i in range(len_window):
+        window_vec[i].first = window[i][0] + 1
+        window_vec[i].second += window[i][1] + 1
+
+    cdef map[pair[int, int], CostEntry] D
+    cdef CostEntry cost_entry
+
+    cost_entry.cost = 0
+    cost_entry.prev_x_idx = 0
+    cost_entry.prev_y_idx = 0
+    inf = float('inf')
+
+    D[__get_key(0, 0)] = cost_entry
+
+    for k in range(len_window):
+        i = window_vec[k].first
+        j = window_vec[k].second
+
+        dt = dist(x[i-1], y[j-1])
```

Comment:
Additionally, I think there is something you could do to C-wrap user-supplied Python distance function by following Cython's callback example: https://github.com/cython/cython/tree/master/Demos/callback

This would allow for supplying a typed function in `cdef _fastdtw`.